x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/NativeMemoryCalculator.java

```diff
@@ -0,0 +1,89 @@
+import static org.elasticsearch.xpack.ml.MachineLearning.MAX_MACHINE_MEMORY_PERCENT;
+import static org.elasticsearch.xpack.ml.MachineLearning.USE_AUTO_MACHINE_MEMORY_PERCENT;
+
+public final class NativeMemoryCalculator {
+
+    private static final long OS_OVERHEAD = ByteSizeValue.ofMb(200L).getBytes();
+
+    private NativeMemoryCalculator() { }
+
+    public static OptionalLong allowedBytesForMl(DiscoveryNode node, Settings settings) {
+        return allowedBytesForMl(
+            node.getAttributes().get(MACHINE_MEMORY_NODE_ATTR),
+            node.getAttributes().get(MAX_JVM_SIZE_NODE_ATTR),
+            MAX_MACHINE_MEMORY_PERCENT.get(settings),
+            USE_AUTO_MACHINE_MEMORY_PERCENT.get(settings));
+    }
+
+    public static OptionalLong allowedBytesForMl(DiscoveryNode node, ClusterSettings settings) {
+        return allowedBytesForMl(
+            node.getAttributes().get(MACHINE_MEMORY_NODE_ATTR),
+            node.getAttributes().get(MAX_JVM_SIZE_NODE_ATTR),
+            settings.get(MAX_MACHINE_MEMORY_PERCENT),
+            settings.get(USE_AUTO_MACHINE_MEMORY_PERCENT));
+    }
+
+    public static OptionalLong allowedBytesForMl(DiscoveryNode node, int maxMemoryPercent, boolean useAuto) {
+        return allowedBytesForMl(
+            node.getAttributes().get(MACHINE_MEMORY_NODE_ATTR),
+            node.getAttributes().get(MAX_JVM_SIZE_NODE_ATTR),
+            maxMemoryPercent,
+            useAuto);
+    }
+
+    private static OptionalLong allowedBytesForMl(String nodeBytes, String jvmBytes, int maxMemoryPercent, boolean useAuto) {
+        if (nodeBytes == null) {
+            return OptionalLong.empty();
+        }
+        final long machineMemory;
+        try {
+            machineMemory = Long.parseLong(nodeBytes);
+        } catch (NumberFormatException e) {
+            return OptionalLong.empty();
+        }
+        Long jvmMemory = null;
+        try {
+            if (Strings.isNullOrEmpty(jvmBytes) == false) {
+                jvmMemory = Long.parseLong(jvmBytes);
+            }
+        } catch (NumberFormatException e) {
+            return OptionalLong.empty();
+        }
+        return OptionalLong.of(allowedBytesForMl(machineMemory, jvmMemory, maxMemoryPercent, useAuto));
+    }
+
+    private static long allowedBytesForMl(long machineMemory, Long jvmSize, int maxMemoryPercent, boolean useAuto) {
+        if (useAuto && jvmSize != null) {
+            // This calculation is dynamic and designed to maximally take advantage of the underlying machine for machine learning
+            // We only allow 200MB for the Operating system itself and take up to 90% of the underlying native memory left
+            // Example calculations:
+            // 1GB node -> 41%
+            // 2GB node -> 66%
+            // 16GB node -> 87%
+            // 64GB node -> 90%
+            long memoryPercent = Math.min(90, (int)Math.ceil(((machineMemory - jvmSize - OS_OVERHEAD) / (double)machineMemory) * 100.0D));
```

Comment:
When this formula was originally used in Cloud we knew the node size would never be less than 200MB.  However, in theory this could now produce a negative number if the node was _really_ small, e.g. in a very constrained Docker container.

Some options would be:

1. Wrap the formula in a `Math.max(1, ...)`
2. Prevent the node starting up if `useAuto` is `true` and node size is less than some value greater than 200MB, e.g. 256MB

Reply:
@droberts195 
> Prevent the node starting up ... node size < 256MB.

I am not sure this is going to be possible. 

It is completely valid to only have this setting set to `true` on `master` nodes. Master nodes don't have to be ML nodes, so our size requirements don't apply to them. It follows that valid ML nodes could have this assigned to `false` and be small, but the master node is set to calculate dynamically.

I see two valid options:

1. do a max calculation as you indicate
2. Fall back to the statically set percentage when the node is too small. 

Option 2 might be unexpected, but I don't know of a nice way to come up with some magic "OS overhead" calculation for exceptionally small nodes. I think exceptionally small nodes just sort of break this calculation. 

I am thinking we should fall back to the static percentage.

Reply:
> I am thinking we should fall back to the static percentage.

I think falling back to 1% would be more reasonable.  Basically we're saying that either there will be an OOM or that significant amounts of swap will need to be used to run a job.  Neither is a very good experience, so if the user specifically asked us to automatically determine how much ML native memory to use then it's reasonable for us to choose an amount that will mean no jobs can get assigned but also nothing will crash or grind to a halt.  The reason for non-assignment will indicate that more memory is needed.